skimage/util/compare.py

```diff
@@ -1,27 +1,143 @@
+import functools
+import inspect
+import warnings
+from itertools import product
+
 import numpy as np
 
 from .dtype import img_as_float
-from itertools import product
+from skimage._shared.utils import (
+    DEPRECATED,
+)
+
+
+class _rename_image_params:
+    """Deprecate parameters `image1, image2` in favour of `image0, image1` in
+    function `compare_images`.
+
+    Parameters
+    ----------
+    deprecated_name : str
+        The name of the deprecated parameter.
+    start_version : str
+        The package version in which the warning was introduced.
+    stop_version : str
+        The package version in which the warning will be replaced by
+        an error / the deprecation is completed.
+    """
+
+    def __init__(
+        self,
+        deprecated_name,
+        *,
+        start_version,
+        stop_version,
+    ):
+        self.deprecated_name = deprecated_name
+        self.start_version = start_version
+        self.stop_version = stop_version
+
+    def __call__(self, func):
+        parameters = inspect.signature(func).parameters
+        if parameters['image2'].default is not DEPRECATED:
+            raise RuntimeError(
+                f"Expected `{self.deprecated_name}` to have the value {DEPRECATED!r} "
+                f"to indicate its status in the rendered signature."
+            )
+        warning_message = (
+            "Since version 0.23, the two input images are named `image0` and "
+            "`image1` (instead of `image1` and `image2`, respectively). Please use "
+            "`image0, image1` to avoid this warning for now, and avoid an error "
+            "from version 0.25 onwards."
+        )
+        wm_method = (
+            "Starting in version 0.25, all arguments following `image0, image1` "
+            "(including `method`) will be keyword-only. Please pass `method=` "
+            "in the function call to avoid this warning for now, and avoid an error "
+            "from version 0.25 onwards."
+        )
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if 'image2' not in kwargs.keys():
+                kwargs['image2'] = DEPRECATED
+
+            # Pass first all args as kwargs
+            if len(args) > 0:
+                kwargs['image0'] = args[0]
+                if len(args) > 1:
+                    kwargs['image1'] = args[1]
+                    if len(args) > 2 and args[len(args) - 1] in [
+                        'diff',
+                        'blend',
+                        'checkerboard',
+                    ]:
```

Comment:
The wrapper shouldn't recognize the `arg` for `method` by value but by position. The function will/should handle if a wrong value is passed as `method`.

Reply:
Yes, the function does handle it:
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L59-L62
Basically I wanted to cover both cases, where a user might call:
```py
compare_images(a, b, 'diff') 
```
and another one might call:
```py
compare_images(a, b, c, 'diff') 
```
which should definitely error -- but this should also be taken care of by the function, right? I realize that my RuntimeError("Use `image0, image1` to pass the two input images.") isn't catching it anyway, it's only catching
```py
compare_images(image0=a, image1=b, image2=c, ...) 
```

Reply:
That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
```python
import skimage as ski
a = np.ones((10, 10)) * 3
b = np.zeros((10, 10))
ski.util.compare_images(a, b, "unknown")
```
This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

Reply:
Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

Reply:
> Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

Right; there's definitely no reason to 'handle' it within the function itself and perhaps not even through the decorator...

Reply:
> That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
> 
> ```python
> import skimage as ski
> a = np.ones((10, 10)) * 3
> b = np.zeros((10, 10))
> ski.util.compare_images(a, b, "unknown")
> ```
> 
> This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

Ouch, yes. Maybe we should also check the args' types then? Let me try something based on your [suggestion](https://github.com/scikit-image/scikit-image/pull/7322/files#r1533795486) and ensure that yet this other edge case is covered.

Reply:
In the end, I was lazy and went for 9b9f757ecc619be (not trying to figure out what the third positional argument was meant to be...).